pallets/evm-accounts/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+			+ Copy
+			+ HasCompact
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ TypeInfo;
+
+		/// EVM nonce provider.
+		type EvmNonceProvider: EvmNonceProvider;
+
+		/// Fee multiplier for the binding of addresses.
+		#[pallet::constant]
+		type FeeMultiplier: Get<u32>;
+
+		/// Weight information for extrinsic in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Maps an EVM address to the last 12 bytes of a substrate account.
+	#[pallet::storage]
+	#[pallet::getter(fn account)]
+	pub(super) type BoundAccount<T: Config> = StorageMap<_, Blake2_128Concat, EvmAddress, AccountIdLast12Bytes>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Binding was created.
+		EvmAccountBounded { who: T::AccountId, evm_address: EvmAddress },
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {
+		/// Nonce is not zero
+		NonZeroNonce,
+		/// Address is already bound
+		AddressAlreadyBound,
+		/// Error occurred while converting AccountId to a different type
+		AccountConversionFailed,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T::AccountId: AsRef<[u8; 32]> + frame_support::traits::IsType<AccountId32>,
+	{
+		/// Binds a Substrate address to EVM address.
+		/// After binding, the EVM is able to convert an EVM address to the original Substrate address.
+		/// Without binding, the EVM converts an EVM address to a truncated Substrate address, which doesn't correspond
+		/// to the origin address.
+		///
+		/// Binding an address is not necessary for interacting with the EVM.
+		///
+		/// Parameters:
+		/// - `origin`: Substrate account binding an address
+		///
+		/// Emits `EvmAccountBound` event when successful.
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::bind_evm_address().saturating_mul(<T as Config>::FeeMultiplier::get() as u64))]
+		pub fn bind_evm_address(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let evm_address = Self::evm_address(&who);
+
+			// This check is not necessary. It prevents binding the same address multiple times.
+			// Without this check binding the address second time can have pass or fail, depending
+			// on the nonce. So it's better to prevent any confusion and throw an error when address is
+			// already bound.
+			ensure!(
+				!BoundAccount::<T>::contains_key(evm_address),
+				Error::<T>::AddressAlreadyBound
+			);
+
+			let nonce = T::EvmNonceProvider::get_nonce(evm_address);
+			ensure!(nonce.is_zero(), Error::<T>::NonZeroNonce);
+
+			let maybe_last_12_bytes: Result<AccountIdLast12Bytes, <[u8; 12] as TryFrom<&[u8]>>::Error> =
+				who.as_ref()[20..32].try_into();
+			let last_12_bytes = if let Ok(bytes) = maybe_last_12_bytes {
+				bytes
+			} else {
+				// this should never happen, but we can't get rid of try_into() when converting a slice to an array
+				return Err(Error::<T>::AccountConversionFailed.into());
+			};
```

Comment:
Thanks. Interesting. Unfortunately it doesn't fully solve the problem. I think it would be better to keep it as it is. Mainly because of that constraint, to prevent integration of the pallet with incompatible types. Wdyt?

Reply:
I am not sure how do you mean it

because you already have the constraint there on line 118.  with that , it will never fail and you can safely use the copy from slice, no need for the try into.

Reply:
actually, i misread the code little bit. i thought initialy that the try into is used to ensure it has 32 bytes too.

but that's not the case, you access the last 12 bytes already and then call try_into.

it is not needed. imo, it does not do anything, yoou will always have the 12 bytes there as you just pass then anyway to try into.

Reply:
try_into is used to convert a slice into a fixed size array